File: src/version6/Messages.js

import Dry from 'json-dry';
import uuidv4 from 'uuid';

import Message from "./Message";
import { KeyError, MissingError } from './Errors';

function _generate_message_uid(){
  let uid = uuidv4();
  return uid.substring(uid.length-8);
}

class Messages {
  constructor(props){
    this.state = {
      registry: {},
    };

    this.load(props);
  };

  add(message){
    if (!(message instanceof Message)){ return;}

    let id = message.getID();

    if (id){
      if (id in this.state.registry){
        throw KeyError(`Duplicate Message ID ${message}`);
      }

      this.state.registry[id] = message;
    }
    else{
      let uid = _generate_message_uid();

      while (uid in this.state.registry){
        uid = _generate_message_uid();
      }

      message.state.id = uid;
      this.state.registry[uid] = message;
    }
  }

  get(id){
    let message = this.state.registry[id];

    if (message === null){
      throw MissingError(`No Message with ID ${id}`);
    }

    return message;
  }

  load(data, people=null){
    if (data){
      if (data.array){ data = data.array; }
      data.forEach(element => {
        let message = Message.load(element, people);
        this.add(message);
      });
    }
  }

  toDry(){
    return {value: this.state.registry};
  }
};

Messages.unDry = function(value){
  let messages = new Messages();
  messages.state.registry = value;
  return messages;
}

Dry.registerClass(Messages);

export default Messages;

File: python/Brunel/_messages.py

import uuid as _uuid
from ._message import Message as _Message

__all__ = ["Messages"]


def _generate_message_uid():
  uid =_uuid.uuid4()
  return "M" + str(uid)[:7]


class Messages:
    """This holds a registry of individual Messages"""
    def __init__(self, props=None):
        self._social = None

        self.state = {
            "registry": {},
        }

        self.load(props)

    def add(self, message: _Message):
        if message is None:
            return

        if not isinstance(message, _Message):
            raise TypeError("Can only add a Message to Messages")

        id = message.getID()

        if id:
            if id in self.state["registry"]:
                raise KeyError(f"Duplicate Message ID {message}")

            self.state["registry"][id] = message
        else:
            uid = _generate_message_uid()

            while uid in self.state["registry"]:
                uid = _generate_message_uid()

            message.state["id"] = uid
            self.state["registry"][uid] = message

        message._social = self._social

    def get(self, id):
        try:
            return self.state["registry"][id]
        except:
            raise KeyError(f"No Message with ID {id}")

    def load(self, data):
        if data:
            for item in data:
                message = _Message.load(item)
                self.add(message)

    def toDry(self):
        return {"value": self.state}

    @staticmethod
    def unDry(value):
        messages = Messages()
        messages.state = value
        return messages
